tools/roxvet/common/traverse.go

```diff
@@ -0,0 +1,62 @@
+package common
+
+import (
+	"go/ast"
+
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+// FilteredPreorder calls `inspector.Preorder(nodeTypes, fn)`, but filters out all files that do not pass the given
+// fileFilter.
+func FilteredPreorder(inspector *inspector.Inspector, fileFilter FileFilter, nodeTypes []ast.Node, fn func(n ast.Node)) {
+	effTypes := nodeTypes[:len(nodeTypes):len(nodeTypes)]
```

Comment:
`eff` = `effective`. Setting the capacity ensures that the call to `append` will allocate a new slice. Too subtle?

Reply:
I had to look up the meaning of the second `:` because I haven't really seen it. Why do we want to ensure a new backing array is allocated? To make `nodeTypes` effectively read-only? Does it have to be read-only?

Reply:
It's just common practice to ensure that parameters are not modified, unless this is made explicit in the function signature.

Additionally, `append` will not modify the visible value of `nodeTypes`. But if somebody passed something like `allNodeTypes[:len(allNodeTypes)-1]`, the last element might get overwritten. This is very subtle and a nightmare to debug, so I'm usually extra careful with `append`.